src/Groups/matrices/form_group.jl

```diff
@@ -0,0 +1,629 @@
+               pos_c +=1
+            end
+            pos_r +=1
+            pos_c =1
+         end
+
+         for i in 1:n^2
+            MM[i,i] -=1
+         end
+         push!(M,MM)
+      end
+
+   else
+
+      w = gen(F) * (gen(F)+gen(F)^q)^-1           # w in F \ F0 s.t. w+w^q=1
+      Nw = em(w*w^q)
+      # we want to write every element of F as a+bw for a,b in F0
+
+      # returns a,b such that x = a+bw
+      coeff2(x) = x+w*(x+x^q), x+x^q
+
+      for mat in gens(G)
+         A = zero_matrix(F0,n,n)
+         B = zero_matrix(F0,n,n)
+         for i in 1:n, j in 1:n           # mat = A + wB,  A,B in GL(n,F0)
+            a,b = coeff2(mat[i,j])
+            A[i,j] = em(a)
+            B[i,j] = em(b)
+         end
+         MM = zero_matrix(F0, n^2, n^2)
+         pos_r = 1
+         pos_c = 1
+
+         # in the vector of solutions, the coefficients are ordered as:
+         # first the x_ii, then the x_ij (i<j) and finally the y_ij (i<j)
+
+         # coefficients for the x_ii
+         for i in 1:n
+            for h in 1:n
+               MM[pos_r,pos_c] = A[i,h]^2+A[i,h]*B[i,h]+Nw*B[i,h]^2
+               pos_c +=1
+            end
+            for h in 1:n, k in h+1:n
+               MM[pos_r,pos_c] = A[i,k]*B[i,h]+A[i,h]*B[i,k]
+               pos_c +=1
+            end
+            for h in 1:n, k in h+1:n
+               MM[pos_r,pos_c] = A[i,h]*A[i,k]+A[i,k]*B[i,h]+Nw*B[i,k]*B[i,h]
+               pos_c +=1
+            end
+            pos_r +=1
+            pos_c =1
+         end
+
+         # coefficients for the x_ij
+         for i in 1:n, j in i+1:n
+            for h in 1:n
+               MM[pos_r,pos_c] = A[i,h]*A[j,h]+A[i,h]*B[j,h]+Nw*B[i,h]*B[j,h]
+               pos_c +=1
+            end
+            for h in 1:n, k in h+1:n
+               MM[pos_r,pos_c] = A[i,h]*A[j,k]+A[i,h]*B[j,k]+A[i,k]*B[j,h]+A[i,k]*A[j,h]+Nw*(B[i,h]*B[j,k]+B[i,k]*B[j,h])
+               pos_c +=1
+            end
+            for h in 1:n, k in h+1:n
+               MM[pos_r,pos_c] = Nw*(A[i,h]*B[j,k]+A[j,k]*B[i,h]+B[i,k]*B[j,h]+A[i,k]*B[j,h]+A[j,h]*B[i,k])+A[i,k]*A[j,h]+A[i,k]*B[j,h]
+               pos_c +=1
+            end
+            pos_r +=1
+            pos_c =1
+         end
+
+         # coefficients for the y_ij
+         for i in 1:n, j in i+1:n
+            for h in 1:n
+               MM[pos_r,pos_c] = A[j,h]*B[i,h]+A[i,h]*B[j,h]
+               pos_c +=1
+            end
+            for h in 1:n, k in h+1:n
+               MM[pos_r,pos_c] = A[j,k]*B[i,h]+A[i,h]*B[j,k]+A[j,h]*B[i,k]+A[i,k]*B[j,h]
+               pos_c +=1
+            end
+            for h in 1:n, k in h+1:n
+               MM[pos_r,pos_c] = A[i,h]*A[j,k]+A[i,k]*A[j,h]+A[j,k]*B[i,h]+A[i,k]*B[j,h]+Nw*(B[i,h]*B[j,k]+B[i,k]*B[j,h])
+               pos_c +=1
+            end
+            pos_r +=1
+            pos_c =1
+         end
+ 
+         for i in 1:n^2
+            MM[i,i] -=1
+         end
+         push!(M,MM)
```

Comment:
This function is ~180 lines long, with no explanation as to what goes on, much less a source :/.

I think I could figure out the details to verify this, but then I dabbled in forms quite a bit in the past. A casual reader of this code will be completely lost here. So it would be great if at the start you could describe the Ansatz being used, in a code comment. The same for the other ,simpler functions (which find quadratic/symmetric/alternating/... forms)

Reply:
I've added some comments.